<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE html>
<html lang="en">
<head>
<meta charset="utf-8"/>
<title>The OrMachine</title>
<meta name="author" content="(Tammo Rukat)"/>

<link rel="stylesheet" href="./reveal.js/css/reveal.css"/>
<link rel="stylesheet" href="./reveal.js/css/theme/sky.css" id="theme"/>

<!-- If the query includes 'print-pdf', include the PDF print sheet -->
<script>
    if( window.location.search.match( /print-pdf/gi ) ) {
        var link = document.createElement( 'link' );
        link.rel = 'stylesheet';
        link.type = 'text/css';
        link.href = './reveal.js/css/print/pdf.css';
        document.getElementsByTagName( 'head' )[0].appendChild( link );
    }
</script>
<script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML,mathjax_local"></script>
</head>
<body>
<div class="reveal">
<div class="slides">
<section id="sec-title-slide" data-background="./logo.png" data-background-size="400px" data-background-repeat="repeat">
<h1>The OrMachine</h1><br><br><br><br><br><h2>Bayesian Boolean matrix factorisation</h2>
</section>


<section>
<section id="slide-sec-1">
<h2 id="sec-1">Single Cell Gene Expression</h2>
<div class="column" style="float:left; width: 100%">

<div class="figure">
<p><img src="./mouse_data.png" alt="mouse_data.png" width="80%" height="80%" />
</p>
</div>
</div>
<aside class="notes">
<p>
Requirements
</p>
<ul class="org-ul">
<li>interpretable: intuition for latent variables, quantification of uncertainty
</li>
<li>correspondencen to physical mechanism: discrete latent properties like presence/absence of disease/mutation
</li>
<li>prior knowledge
</li>
<li>scalability
</li>
</ul>

</aside>

</section>
</section>
<section>
<section id="slide-sec-2">
<h2 id="sec-2">Boolean Matrix Factorisation</h2>
<div class="column" style="float:left; width: 50%">
<p>
$$ $$
</p>
<span class="fragment (appear)" data-fragment-index="1"><p>
<p>
Observed Data
</p>

<div class="figure">
<p><img src="./calc_digit_data.png" alt="calc_digit_data.png" data-fragment-index="1" class="fragment appear" width="90%" height="90%" />
</p>
</div>
</div>

<div class="column" style="float:left; width: 50%">
<span class="fragment (appear)" data-fragment-index="2"><p>
<p>
Factorisation
</p>

<div class="figure">
<p><img src="./calc_digit_factor.png" alt="calc_digit_factor.png" data-fragment-index="2" class="fragment appear" width="90%" height="90%" />
</p>
</div>
</div>

<div class="column" style="float:left; width: 100%">
<span class="fragment (appear)" data-fragment-index="3"><p>
<p>
Example
</p>

<div class="figure">
<p><img src="./calc_example.png" alt="calc_example.png" data-fragment-index="3" class="fragment appear)" width="45%" height="45%" />
</p>
</div>
</div>

<aside class="notes">
<ul class="org-ul">
<li>What is BooMF? The approximation of a binary data matrix as product of two low rank binary matrices.
</li>
<li>Like a regular matrix product but thresholded.
</li>
<li>Essentially the model learns the seperate bars from which every digit can be composed. We call them <i>codes</i>
</li>
<li>The L-dimensional indicator Z provides the compact representation of which codes are allocated to each observation.
</li>
<li>Similarity to noisy-OR
</li>
</ul>

</aside>

</section>
<section id="slide-sec-2-1">
<h3 id="sec-2-1">Probabilistic Generative Model</h3>
<div class="column" style="float:right; width: 100%">
<span class="fragment appear"><p>
$$  p(\underbrace{x_{nd}}_{\substack{\text{obser-} \\ \text{vation}}}|\overbrace{\mathbf{u}_d}^{\text{codes}},\underbrace{\mathbf{z}_n}_{\substack{\text{latent}\\ \text{rprsnt.}}},\overbrace{\lambda}^{\substack{\text{disper-}\\ \text{sion}}})= \begin{cases} \big(1+\exp[-\lambda]\big)^{-1};\;&\text{if}\;\color{darkgreen}{x_{nd}=\min(1,\mathbf{z}_n^T\mathbf{u}_d)}\;\; \\ \big(1+\exp[\lambda]\big)^{-1};\;&\text{else} \end{cases}
$$
</p></span>
<span class="fragment appear"><p>
$$\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\; = \sigma_{\substack{\text{logistic} \\ \text{sigmoid}}}\left[\lambda \underbrace{\tilde{x}_{nd}}_{\tilde{x} = 2x-1} \left(1-2\color{brown}{\prod\limits_{l}(1-z_{nl}u_{ld})}\right) \right]$$
</p></span>
</div>
<aside class="notes">
<ul class="org-ul">
<li>This is a probabilistic model for <b>Boolean matrix factorisation</b>.
</li>
<li>Likelihood can be efficiently evaluated
</li>
<li>Example topic models
</li>
</ul>

</aside>
</section>
</section>
<section>
<section id="slide-sec-3">
<h2 id="sec-3">Inference for the OrMachine</h2>
<p>
<b>Gibbs Sampling</b>
</p>
<div class="column" style="float:left; width: 100%">
<div class="column" style="float:left; width: 60%">
<span class="fragment (appear)" data-fragment-index="1"><p>
<p>
Full conditional
</p>
<span class="fragment (appear)" data-fragment-index="1"><p>
<p>
$$ p(z_{nl}|\text{rest}) = \sigma\bigg[\lambda \tilde{z}_{nl} \sum\limits_d \tilde{x}_{nd}\; \color{darkgreen}{u_{ld}} \color{brown}{\prod\limits_{l'\neq l} (1-z_{nl'}u_{l'd})}\bigg] $$
</p>
</div>

<div class="column" style="float:left; width: 35%">
<span class="fragment (appear)" data-fragment-index="2"><p>
<p>
Computational shortcuts
</p>
<ul>
<li>\(\color{darkgreen}{u_{ld} = 0}\) <br> &rarr; \(z_{nl}\) and \(x_{nd}\) <i>disconnected</i>.</li>
<li>\(\color{brown}{z_{nl'}u_{l'd} = 1}\) for \(\color{brown}{l' \neq l}\) <br> &rarr; \(x_{nd}\) is <i>explained away</i>.</li>

</ul>
</div>

<span class="fragment (appear)" data-fragment-index="3"><p>
<p>
<b>Dispersion parameter set to maximise likelihood</b>
</p>
<span class="fragment (appear)" data-fragment-index="3"><p>
<p>
$$ \lambda_{\text{MLE}} = \text{logit}\left[ \text{reconstruction accuracy} \right] $$
</p>
</div>

<div class="column" style="float:left; width: 100%">
<span class="fragment (appear)" data-fragment-index="4"><p>
<p>
<b>Modified Sampler &#x2013; Always propose to change</b>
</p>
<div class="column" style="float:left; width: 50%">
<span class="fragment (appear)" data-fragment-index="4"><p>
<p>
Standard Gibbs sampler <br> 
  <img src="./heads_smaller.png" alt="heads_smaller.png" />   <img src="./heads_smaller.png" alt="heads_smaller.png" />  <img src="./heads_smaller.png" alt="heads_smaller.png" />  <img src="./tails_smaller.png" alt="tails_smaller.png" />  <img src="./heads_smaller.png" alt="heads_smaller.png" /> 
 <img src="./tails_smaller.png" alt="tails_smaller.png" />  <img src="./tails_smaller.png" alt="tails_smaller.png" />  <img src="./heads_smaller.png" alt="heads_smaller.png" />  <img src="./tails_smaller.png" alt="tails_smaller.png" /> 
</p>
</div>
<div class="column" style="float:left; width: 50%">
<span class="fragment (appear)" data-fragment-index="4"><p>
<p>
Metropolised Gibbs sampler <br> 
  <img src="./heads_smaller.png" alt="heads_smaller.png" />  <img src="./tails_smaller.png" alt="tails_smaller.png" />   <img src="./heads_smaller.png" alt="heads_smaller.png" />  <img src="./tails_smaller.png" alt="tails_smaller.png" />  <img src="./heads_smaller.png" alt="heads_smaller.png" />  <img src="./tails_smaller.png" alt="tails_smaller.png" />  <img src="./heads_smaller.png" alt="heads_smaller.png" />  <img src="./tails_smaller.png" alt="tails_smaller.png" />  <img src="./heads_smaller.png" alt="heads_smaller.png" />
</p>
</div>


<aside class="notes">
<ul class="org-ul">
<li>Use graphical model language -&gt; prior knowledge is clamping variables
</li>
<li>Conditional takes a surprisingly simple form
</li>
<li>You might think this is slow
</li>
<li>Think if this as a neural net without nonlinearity
</li>
<li>Normally need to consider <b>full Markov blanket</b>
</li>
<li>Parallelisable
</li>
<li>Set lambda to MLE after every sweep through u/z in <b>Monte Carlo EM fashion</b>
</li>
<li>Parallelisable
</li>
</ul>

</aside>

</section>
</section>
<section>
<section id="slide-sec-4">
<h2 id="sec-4">Examples and Experiments</h2>
<div class="outline-text-2" id="text-4">
</div></section>
<section id="slide-sec-4-1">
<h3 id="sec-4-1">Synthetic Data Benchmarks</h3>
<div class="column" style="float:left; width: 50%">
<span class="fragment (appear)" data-fragment-index="1"><p>
<p>
<b>Random Matrix Factorisation</b>
</p>

<div class="figure">
<p><img src="./factorsiation_performance_new2.png" alt="factorsiation_performance_new2.png" data-fragment-index="1" class="fragment appear" width="75%" height="%0%" />
</p>
</div>

<p>
[Message Passing: Ravanbakshs et al., ICML 2016]
</p>
</div>

<div class="column" style="float:left; width: 50%">
<span class="fragment (appear)" data-fragment-index="1"><p>
<p>
Problem Setting
</p>

<div class="figure">
<p><img src="./factor_example2.png" alt="factor_example2.png" data-fragment-index="1" class="fragment appear" width="95%" height="95%" />
</p>
</div>



<span class="fragment (appear)" data-fragment-index="2"><p>
<p>
<b>Random Matrix Completion</b>
</p>

<div class="figure">
<p><img src="./completion1.png" alt="completion1.png" data-fragment-index="2" class="fragment appear" width="60%" height="60%" />
</p>
</div>
</div>




<aside class="notes">
<ul class="org-ul">
<li>Compare to state of the art
</li>
<li>Movie lense results are in the paper
</li>
</ul>

</aside>
</section>
<section id="slide-sec-4-2">
<h3 id="sec-4-2">Single cell data II - 1.3 Million Brain Cells x 20k genes (E18 Mice)</h3>
<span class="fragment (appear)" data-fragment-index="1"><p>
<p>
<img src="./mouse_gene_single.png" alt="mouse_gene_single.png" width="110%" height="110%" />
Codes: Gene sets
</p>
<span class="fragment (appear)" data-fragment-index="1"><p>
<p>
<img src="./mouse_specimen_single.png" alt="mouse_specimen_single.png" width="110%" height="110%" />
Latent representations of each cell
$$ $$
</p>
<span class="fragment (appear)"><p>
<b>How to chose the latent dimension?</b>
</p></span>

<aside class="notes">
<ul class="org-ul">
<li>Convergence in a few hours on a Desktop with 8 cores. <b>Parallelsiable</b>!
</li>
</ul>

</aside>

</section>
<section id="slide-sec-4-2-1">
<h4 id="sec-4-2-1">Calculator Digit Hierarchy</h4>
<div class="column" style="float:left; width: 100%">

<div class="figure">
<p><img src="./calc_hierarchy_may5.png" alt="calc_hierarchy_may5.png" />
</p>
</div>
<ul>
<li>OrMachines of different dimensionality on noise-free calculator digits</li>

</ul>
</div>
</section>
<section id="slide-sec-4-2-2">
<h4 id="sec-4-2-2">Gene patterns &#x2013; Cell representations</h4>
<div class="column" style="float:left; width: 100%">

<div class="figure">
<p><img src="./mice_neurons1.png" alt="mice_neurons1.png" width="73%" height="50%" />
</p>
</div>

<div class="figure">
<p><img src="./mice_neurons2.png" alt="mice_neurons2.png" data-fragment-index="1" class="fragment appear" width="73%" height="50%" />
</p>
</div>
</div>

</section>
<section id="slide-sec-4-3">
<h3 id="sec-4-3">Mutations and Cellular Pathways in Different Cancer Types</h3>

</section>
<section id="slide-sec-4-4">
<h3 id="sec-4-4">Compose Layers of OrMachines</h3>
<div class="column" style="float:left; width: 19%">

<div class="figure">
<p><img src="./compose_mutations.png" alt="compose_mutations.png" width="100%" height="100%" /> 
</p>
</div>
</div>
<div class="column" style="float:left; width: 3%">
<p>
$$ $$
$$ $$
$$ $$
\(\mathbf{\otimes}\)
</p>
</div>
<div class="column" style="float:left; width: 19%">
<p>
<img src="./compose_pws.png" alt="compose_pws.png" width="100%" height="100%" />
$$ $$
$$ $$
</p>
<span class="fragment (appear)" data-fragment-index="3"><p>

<div class="figure">
<p><img src="./compose_disease.png" alt="compose_disease.png" width="100%" height="100%" />
</p>
</div>
</div>
<div class="column" style="float:left; width: 5%">
<p>
$$ $$
$$ $$
$$ $$
</p>
<span class="fragment (appear)" data-fragment-index="1"><p>
<p>
\(\mathbf{\;\rightarrow\;}\)
$$ $$
$$ $$
$$ $$
$$ $$
$$ $$
$$ $$
$$ $$
$$ $$
</p>
<span class="fragment (appear)" data-fragment-index="3"><p>
<p>
\(\mathbf{\;\;\;\;\otimes}\)
</p>
</div>
<div class="column" style="float:left; width: 25%">
<span class="fragment (appear)" data-fragment-index="1"><p>
<p>
<img src="./compose_pw_patients_empty.png" alt="compose_pw_patients_empty.png" width="100%" height="100%" />
$$ $$
$$ $$
</p>
<span class="fragment (appear)" data-fragment-index="3"><p>

<div class="figure">
<p><img src="./compose_set_sets_empty.png" alt="compose_set_sets_empty.png" width="75%" height="75%" />
</p>
</div>
</div>
<div class="column" style="float:left; width: 5%">
<p>
$$ $$
$$ $$
$$ $$
</p>
<span class="fragment (appear)" data-fragment-index="2"><p>
<p>
\(\mathbf{\;\leftarrow}\)
$$ $$
$$ $$
$$ $$
$$ $$
$$ $$
$$ $$
$$ $$
$$ $$
</p>
<span class="fragment (appear)" data-fragment-index="3"><p>
<p>
\(\mathbf{\rightarrow\;\;}\)
</p>
</div>

<div class="column" style="float:left; width: 23%">
<span class="fragment (appear)" data-fragment-index="2"><p>
<p>
<img src="./compose_pw_sets_empty.png" alt="compose_pw_sets_empty.png" width="85%" height="85%" />
<br>
</p>
<span class="fragment (appear)" data-fragment-index="2"><p>
<p>
\(\mathbf{\otimes}\)
</p>
<span class="fragment (appear)" data-fragment-index="2"><p>

<div class="figure">
<p><img src="./compose_embedding_empty.png" alt="compose_embedding_empty.png" width="85%" height="85%" /> 
</p>
</div>
</div>

<aside class="notes">
<ul class="org-ul">
<li><b>So this illustrates how this approach infers biologically meaningful differences and communalities between and within patient groups, that can lead to testable hypothesis, rather than optimising for classificatoin or prediction performance as continuous representations would.</b> 
</li>

<li>Discrete latent variables lead to testable hypothesis rather than classification or prediction performance 
</li>
<li>From a Bayesian perspective this is like a prior, indicating more factorisable structure.
</li>
<li>This expresses the prior belief that patients with the same disease share latent properties.
</li>
</ul>

</aside>

</section>
<section id="slide-sec-4-4-1" data-transition="none">
<h4 id="sec-4-4-1">Compose Layers of OrMachines</h4>
<div class="column" style="float:left; width: 19%">

<div class="figure">
<p><img src="./compose_mutations.png" alt="compose_mutations.png" width="100%" height="100%" />
</p>
</div>
</div>
<div class="column" style="float:left; width: 3%">
<p>
$$ $$
$$ $$
$$ $$
\(\mathbf{\otimes}\)
</p>
</div>
<div class="column" style="float:left; width: 19%">
<p>
<img src="./compose_pws.png" alt="compose_pws.png" width="100%" height="100%" />
$$ $$
$$ $$
</p>

<div class="figure">
<p><img src="./compose_disease.png" alt="compose_disease.png" width="100%" height="100%" />
</p>
</div>
</div>
<div class="column" style="float:left; width: 5%">
<p>
$$ $$
$$ $$
$$ $$
\(\mathbf{\;\rightarrow\;}\)
$$ $$
$$ $$
$$ $$
$$ $$
$$ $$
$$ $$
$$ $$
$$ $$
\(\mathbf{\;\;\;\;\otimes}\)
</p>
</div>
<div class="column" style="float:left; width: 25%">
<p>
<img src="./compose_pw_patients.png" alt="compose_pw_patients.png" width="100%" height="100%" />
$$ $$
$$ $$
</p>

<div class="figure">
<p><img src="./compose_set_sets.png" alt="compose_set_sets.png" width="75%" height="75%" />
</p>
</div>
</div>
<div class="column" style="float:left; width: 5%">
<p>
$$ $$
$$ $$
$$ $$
\(\mathbf{\;\leftarrow}\)
$$ $$
$$ $$
$$ $$
$$ $$
$$ $$
$$ $$
$$ $$
$$ $$
\(\mathbf{\rightarrow\;\;}\)
</p>
</div>

<div class="column" style="float:left; width: 23%">
<p>
<img src="./compose_pw_sets.png" alt="compose_pw_sets.png" width="85%" height="85%" />
<br>
\(\mathbf{\otimes}\)
</p>

<div class="figure">
<p><img src="./compose_embedding.png" alt="compose_embedding.png" width="85%" height="85%" /> 
</p>
</div>
</div>

</section>
<section id="slide-sec-4-4-2">
<h4 id="sec-4-4-2">Embedding</h4>
<div class="column" style="float:left; width: 100%">

<div class="figure">
<p><img src="./pancan_representations.png" alt="pancan_representations.png" width="75%" height="75%" />
</p>
</div>
</div>
</section>
</section>
<section>
<section id="slide-sec-5">
<h2 id="sec-5">Conclusion</h2>
<div class="column" style="float:left; width: 100%">
<ul>
<li data-fragment-index="1" class="fragment appear">Outperforms available methods for Boolean Matrix Factorisation</li>
<li data-fragment-index="2" class="fragment appear">Applicable and scalable to most state-of-the art genomics data.</li>
<li data-fragment-index="3" class="fragment appear">Compose OrMachines to link different types of knowledge.</li>

</ul>
</div>
</section>
<section id="slide-sec-5-1">
<h3 id="sec-5-1">Acknowledgements</h3>
<p>
<b>Supervision</b>
</p>

<div class="column" style="float:left; width: 20%">
<p>
$$ $$
</p>
</div>

<div class="column" style="float:left; width: 20%">
<p>
<img src="./chris.jpg" alt="chris.jpg" width="65%" height="65%" /> <br>
Chris Yau
</p>
</div>

<div class="column" style="float:left; width: 20%">
<p>
<img src="./michalis.png" alt="michalis.png" width="65%" height="65%" /> <br>
Michalis Titsias
</p>
</div>

<div class="column" style="float:left; width: 20%">
<p>
<img src="./ch2.jpg" alt="ch2.jpg" width="60%" height="60%" /> <br>
Chris Holmes
</p>
</div>

<div class="column" style="float:left; width: 50%">
<span class="fragment (appear)" data-fragment-index="1"><p>
<p>
<b>Funding</b>
</p>
</div>
<div class="column" style="float:left; width: 50%">
<span class="fragment (appear)" data-fragment-index="2"><p>
<p>
<b>Find me next academic year @</b>
</p>
</div>
<div class="column" style="float:left; width: 25%">
<span class="fragment (appear)" data-fragment-index="1"><p>

<div class="figure">
<p><img src="./epsrc.png" alt="epsrc.png" width="70%" height="80%" />
</p>
</div>
</div>
<div class="column" style="float:left; width: 25%">
<span class="fragment (appear)" data-fragment-index="1"><p>

<div class="figure">
<p><img src="./roche.png" alt="roche.png" width="70%" height="80%" />
</p>
</div>
</div>
<div class="column" style="float:left; width: 50%">
<span class="fragment (appear)" data-fragment-index="2"><p>

<div class="figure">
<p><img src="./ait.png" alt="ait.png" width="32%" height="32%" />
</p>
</div>
</div>
</section>
</section>
</div>
</div>

<script src="./reveal.js/lib/js/head.min.js"></script>
<script src="./reveal.js/js/reveal.js"></script>

<script>
// Full list of configuration options available here:
// https://github.com/hakimel/reveal.js#configuration
Reveal.initialize({

controls: true,
progress: true,
history: false,
center: true,
slideNumber: true,
rollingLinks: true,
keyboard: true,
overview: true,
width: 1920,
height: 1080,
margin: 0.15,
minScale: 0.50,
maxScale: 2.00,

theme: Reveal.getQueryHash().theme, // available themes are in /css/theme
transition: Reveal.getQueryHash().transition || 'cube', // default/cube/page/concave/zoom/linear/fade/none
transitionSpeed: 'default',
multiplex: {
    secret: '', // null if client
    id: '', // id, obtained from socket.io server
    url: '' // Location of socket.io server
},

// Optional libraries used to extend on reveal.js
dependencies: [
 { src: './reveal.js/plugin/notes/notes.js', async: true, condition: function() { return !!document.body.classList; } },
 { src: './reveal.js/plugin/highlight/highlight.js', async: true, callback: function() { hljs.initHighlightingOnLoad(); } },
 { src: './reveal.js/plugin/markdown/marked.js', condition: function() { return !!document.querySelector( '[data-markdown]' ); } },
 { src: './reveal.js/plugin/markdown/markdown.js', condition: function() { return !!document.querySelector( '[data-markdown]' ); } }
]
});
</script>
</body>
</html>
